### PR TITLE
fix(gatsby): handle initializing multiple instances of gatsby-plugin-sharp

### PR DESCRIPTION
**Pull request by pieh**
## Description

Fix for

```
error Gatsby-plugin-sharp wasn't setup correctly in gatsby-config.js. Make sure you add it to the plugins array.
```

errors when there's multiple instances of gatsby-plugin-sharp package (for themes scenarios, when there might be different versions and package manager doesn't or can't dedupe)

Explanation of changes:
1. We where deduping plugins when merging themes, which would be fine, but at that point we don't know yet what those plugins resolve to so we were potentially throwing away plugin entries that seemed the same but actually weren't. Case of `gatsby-plugin-sharp` which most often doesn't have any options and was deduped because it's specified just as `gatsby-plugin-sharp` - this meant that some instances of `gatsby-plugin-sharp` were not initiated at all. This PR moves it to after we resolve plugins where we can check actual plugin packages instances when deduping. This also required update to `import-gatsby-plugin` to be able to not just use plugin name, but different identifier - for `gatsby build/develop` we rely on `plugin.resolve` property. I did have to keep fallback to `name` for `gatsby-worker`, as when trying to use it we are losing which instance it is for.
2. Engines needed adjustments to not use plugin's name for dictionary keys as then we would lose information when there is multiple instances left after deduping, instead we use array. Additionally as we are dropping/clearing `resolve` property from engines I needed yet another way to adjust `import-gatsby-plugin` and used simple `importKey` which is 1-indexed counter (and it represent just the index of plugin in the plugins list
3. Used this opportunity to convert `load-themes/index.js` to typescript

### CODE REVIEW

File: packages/gatsby/src/schema/graphql-engine/standalone-regenerate.ts [hunk 1]
```diff
@@ -26,6 +26,7 @@ import { store } from "../../redux"
 import { validateEngines } from "../../utils/validate-engines"
 
 async function run(): Promise<void> {
+  process.env.GATSBY_SLICES = `1`
```
Comment (tyhopp) [inline packages/gatsby/src/schema/graphql-engine/standalone-regenerate.ts:29]: Double checking, is this change also required for the fix or left in accidentally?

**Review by tyhopp — COMMENTED** [review]
Looks great, one question and then I'll approve!

File: packages/gatsby/src/schema/graphql-engine/standalone-regenerate.ts (hunk 1 above)
Comment (pieh) [reply packages/gatsby/src/schema/graphql-engine/standalone-regenerate.ts:29]: This is not needed for a fix as we don't actually use this file in regular usage. 

I just noticed problems with trying to use engines regenerated with this script manually as I was iterating on this PR and testing produced engines

**Review by tyhopp — APPROVED** [review]

**Comment by tyhopp** [comment]
Published in `gatsby@4.25.2`

**Pull request merged**